concept/type_/annotation.rs

```diff
@@ -115,6 +162,262 @@ impl AnnotationRegex {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        hash_value::hash_value_opt(&self.start_inclusive, state);
+        hash_value::hash_value_opt(&self.end_inclusive, state);
+    }
+}
+
+#[derive(Debug, Default, Clone, Eq, PartialEq)]
+pub struct AnnotationValues {
+    // ##########################################################################
+    // ###### WARNING: any changes here may break backwards compatibility! ######
+    // ##########################################################################
+    values: Vec<Value<'static>>,
+}
+
+impl AnnotationValues {
+    pub const fn new(values: Vec<Value<'static>>) -> Self {
+        Self { values }
+    }
+
+    pub const fn default() -> Self {
+        Self { values: vec![] }
+    }
+
+    pub fn values(&self) -> &[Value<'static>] {
+        &self.values
+    }
+
+    // TODO: We might want to return different errors for empty / unmatched value types
+    pub fn valid(&self, expected_value_type: Option<ValueType>) -> bool {
+        if self.values.is_empty() {
+            return false;
+        }
+
+        let unique_value_types: HashSet<ValueType> = self.values.iter().map(|value| value.value_type()).collect();
+        assert!(!unique_value_types.is_empty());
+        if unique_value_types.len() > 1 {
+            return false;
+        }
+        if expected_value_type.is_some()
+            && unique_value_types.iter().any(|value_type| value_type != &expected_value_type.clone().unwrap())
+        {
+            return false;
+        }
+
+        // Value does not implement Hash, so we run a N^2 loop here expecting a limited number of values
+        let values = &self.values;
+        for i in 0..values.len() {
+            for j in i + 1..values.len() {
+                if values[i] == values[j] {
+                    return false;
+                }
+            }
+        }
+
+        true
+    }
+
+    pub fn value_valid(&self, value: Value<'static>) -> bool {
+        self.contains(&value)
+    }
+
+    pub fn value_type_valid(value_type: Option<ValueType>) -> bool {
+        match value_type {
+            Some(value_type) => match &value_type {
+                | ValueType::Boolean
+                | ValueType::Long
+                | ValueType::Double
+                | ValueType::Decimal
+                | ValueType::Date
+                | ValueType::DateTime
+                | ValueType::DateTimeTZ
+                | ValueType::Duration
+                | ValueType::String => true,
+
+                | ValueType::Struct(_) => false,
+            },
+            None => false,
+        }
+    }
+
+    pub fn narrowed_correctly_by(&self, other: &Self) -> bool {
+        for other_value in &other.values {
+            if !self.contains(&other_value) {
+                return false;
+            }
+        }
+        true
+    }
+
+    fn contains(&self, value: &Value<'static>) -> bool {
+        self.values.contains(value)
+    }
+}
+
+impl Hash for AnnotationValues {
```

Comment:
We will be able to use `derive` here after we sync branches! I'll control that we don't miss it